### PR TITLE
Improve spectral_norm (fixes #7261)

### LINKED ISSUE #7261: Possible memory leak in spectral_norm
I wanted to give a try to spectral_norm function recently included in master through commit ba046331e8cfae7c93a86a5664fcb5c25f9dbee0

I observe that the same model with spectral_norm() around Conv2d & Linear layers keeps its GPU memory increasing until OOM exception.

Removing spectral_norm() call makes everything run smoothly.

- PyTorch: master
- How you installed PyTorch (conda, pip, source): source
- Build command you used (if compiling from source): python setup.py install
- OS: CentOS Linux release 7.4.1708 (Core)
- PyTorch version: master c96f2624a280ae3e2d9195c35150f6aec85f6a02
- Python version: 3.6.5
- CUDA/cuDNN version: 9.1 / 7.1.3
- GPU models and configuration: V100 x 4
- GCC version (if compiling from source): 4.8
- CMake version: 3.5

### PR DESCRIPTION
**Pull request by t-vi**
Thank you Morgan Funtowicz for the report and minimal example!

This eliminates the memory leaking behavious that Morgan observed insofar as `print ("memory", torch.cuda.memory_allocated())` is stable (with some noise - i.e. oscillations between nearby values, but not increasing).

In the course of the above, it makes two functional changes to the code: I assume that we do *not* need gradients for the spectral of the `weight` nor `u`, but only for the `weight` and by the rescaled weight. On the other hand, I divide the weight rather than weight.data to not get into trouble with the backward. This brings the code closer to my interpretation of the reference. This should probably be double-checked who knows bettter (@crcrpar)?

### CODE REVIEW

File: torch/nn/utils/spectral_norm.py [hunk 1]
```diff
@@ -14,32 +14,34 @@ def __init__(self, name='weight', n_power_iterations=1, eps=1e-12):
...
-        for _ in range(self.n_power_iterations):
-            # Spectral norm of weight equals to `u^T W v`, where `u` and `v`
-            # are the first left and right singular vectors.
-            # This power iteration produces approximations of `u` and `v`.
-            v = normalize(torch.matmul(weight_mat.t(), u), dim=0, eps=self.eps)
-            u = normalize(torch.matmul(weight_mat, v), dim=0, eps=self.eps)
-
-        sigma = torch.dot(u, torch.matmul(weight_mat, v))
-        weight.data /= sigma
+        with torch.no_grad():
```
Comment (ezyang) [inline torch/nn/utils/spectral_norm.py]: Yes please :)

**Comment by mfuntowicz** [comment]
Just cherry-picked your commit and I can confirm that there is no leak anymore :).

Thank for fixing so quickly @t-vi !
Morgan

**Comment by ezyang** [comment]
For my curiosity, why does switching to getattr/setattr solve the leakage?

**Comment by crcrpar** [comment]
Thank you for fixing @t-vi !
I'm sorry @mfuntowicz

**Comment by t-vi** [comment]
I think the key fix is using no_grad and .copy_. The *attr is to use more of the official interface rather than internals.

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by t-vi** [comment]
@crcpar He, thank you for your implementation!
I wonder if the technical python trickery should be wrapped into a proper abstraction - another use case for this type of "precomputation" would be transformed / constrained parameters.

File: torch/nn/utils/spectral_norm.py [hunk 2]
```diff
@@ -14,32 +14,34 @@ def __init__(self, name='weight', n_power_iterations=1, eps=1e-12):
...
+            for _ in range(self.n_power_iterations):
+                # Spectral norm of weight equals to `u^T W v`, where `u` and `v`
+                # are the first left and right singular vectors.
+                # This power iteration produces approximations of `u` and `v`.
+                v = normalize(torch.matmul(weight_mat.t(), u), dim=0, eps=self.eps)
+                u = normalize(torch.matmul(weight_mat, v), dim=0, eps=self.eps)
+
+                sigma = torch.dot(u, torch.matmul(weight_mat, v))
```
Comment (fmassa) [inline torch/nn/utils/spectral_norm.py]: Do we need to compute the sigma in the loop? It's different from before

Comment (t-vi) [reply torch/nn/utils/spectral_norm.py]: No, good catch, thank you!

**Review by ezyang — APPROVED** [review]

File: torch/nn/utils/spectral_norm.py [hunk 3]
```diff
@@ -14,32 +14,34 @@ def __init__(self, name='weight', n_power_iterations=1, eps=1e-12):
...
-        weight.data /= sigma
+        with torch.no_grad():
+            for _ in range(self.n_power_iterations):
+                # Spectral norm of weight equals to `u^T W v`, where `u` and `v`
+                # are the first left and right singular vectors.
+                # This power iteration produces approximations of `u` and `v`.
+                v = normalize(torch.matmul(weight_mat.t(), u), dim=0, eps=self.eps)
+                u = normalize(torch.matmul(weight_mat, v), dim=0, eps=self.eps)
+
+            sigma = torch.dot(u, torch.matmul(weight_mat, v))
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: This should be outside `no_grad`...

Comment (t-vi) [reply torch/nn/utils/spectral_norm.py]: Are you sure? I don't think sigma should be backpropped into (u and v are defined by the iterative procedure, not gradient descent).

Comment (SsnL) [reply torch/nn/utils/spectral_norm.py]: Yes, see sec 2.3 of https://arxiv.org/pdf/1802.05957.pdf. u and v are not backprop'ed but weight_mat is. I'm working on a fix.
